waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -82,22 +79,34 @@ data TypeCoercionErrorReason e
     ReasonDictWrongKeyType String e
   deriving (Eq, Show)
 
-getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
-getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
+extractTypeErrorMessagesAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, [String], Ctx)
+extractTypeErrorMessagesAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
   case reason of
     ReasonList e ->
-      first (("For list element:\n" ++) . indent 2) $
-        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+      second ("-> In list element" :) $ extractTypeErrorMessagesAndCtx getUncoercableTypesMsg e
     ReasonDictWrongKeyType key e ->
-      first ((("For dictionary field '" ++ key ++ "':\n") ++) . indent 2) $
-        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
-    ReasonDictNoKey key -> ("Missing required dictionary field '" ++ key ++ "'", ctx)
-    ReasonDictExtraKey key -> ("Unexpected dictionary field '" ++ key ++ "'", ctx)
+      second (("-> For dictionary field '" ++ key ++ "'") :) $ extractTypeErrorMessagesAndCtx getUncoercableTypesMsg e
+    ReasonDictNoKey key -> ("Missing required dictionary field '" ++ key ++ "'", [], ctx)
+    ReasonDictExtraKey key -> ("Unexpected dictionary field '" ++ key ++ "'", [], ctx)
     ReasonDecl -> uncoercableTypesMsgAndCtx
     ReasonEnum -> uncoercableTypesMsgAndCtx
     ReasonUncoercable -> uncoercableTypesMsgAndCtx
   where
-    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, ctx)
+    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, [], ctx)
+
+joinAdditionalMessages :: [String] -> String
+joinAdditionalMessages [] = ""
+joinAdditionalMessages msgChain = foldr1 appendMsg msgChain
+  where
+    appendMsg curr acc = intercalate ":\n" [curr, indent 2 acc]
+
+getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
```

Comment:
In general, the order of functions in this file is not great. Normally we try to keep them ordered top -> bottom, so that the ones above use the others below.
This is not your doing, they were already ordered wrong, by me previously.
But we could fix that now. So `getUnificationErrorMessageAndCtx` and `getWeakenErrorMessageAndCtx` would go first, then we would have `getTypeCoercionErrorMessageAndCtx`, and only then we would have `extractTypeErrorMessagesAndCtx`, so it reads like a story, first high level stuff, then the implementation details.